### PR TITLE
Only use v8 serialization internally within the thunk

### PR DESCRIPTION
**Pull request by dweitzman-codaio**
_**This is a draft PR, not ready for review yet**_

### CODE REVIEW

File: runtime/common/marshaling/index.ts [hunk 1]
```diff
@@ -173,12 +174,17 @@ export function marshalValue(val: any): MarshaledValue {
...
-export function marshalValueToString(val: any): string {
+export function marshalValueToStringForSameOrHigherNodeVersion(val: any): string {
   return serialize(marshalValue(val));
 }
 
 export function unmarshalValueFromString(marshaledValue: string): any {
-  return unmarshalValue(deserialize(marshaledValue));
+  if (marshaledValue?.startsWith('/')) {
+    // Looks like a v8-serialized value
+    return unmarshalValue(deserialize(marshaledValue));
```
Comment (chris-codaio) [inline runtime/common/marshaling/index.ts:184]: Maybe a try/catch around `deserialize` in case that fails?

Comment (dweitzman-codaio) [reply runtime/common/marshaling/index.ts:250]: What error handling were you thinking here?

I don't think we'd want to suppress any errors here for fear of moving incorrect data around. This can't be a legacy marshaled value if it starts with "/" since that can't be the beginning of JSON string.

Comment (chris-codaio) [reply runtime/common/marshaling/index.ts:250]: Ah, I suppose that's sufficient then.

File: helpers/legacy_marshal/index.ts [hunk 2]
```diff
@@ -91,7 +91,7 @@ function processValue(val: any, depth: number = 0): any {
   return serializedValue;
 }
 
-export function legacyMarshalValue(val: any): string | undefined {
+export function marshalValueForAnyNodeVersion(val: any): string {
```
Comment (jonathan-codaio) [inline helpers/legacy_marshal/index.ts:94]: Do you want to rename the unmarshaler too?

File: test/test_helper.ts [hunk 3]
```diff
@@ -6,9 +6,13 @@ export const testHelper = {
    * @param promise The original promise.
    * @returns The inverted promise, which fulfills if rejected and rejects if fulfilled.
    */
-  async willBeRejected<ErrorT = any>(promise: Promise<any>): Promise<ErrorT> {
+  async willBeRejected<ErrorT = any>(promise: Promise<any> | (() => Promise<any>)): Promise<ErrorT> {
     try {
-      await promise;
+      if (typeof promise === 'function') {
```
Comment (jonathan-codaio) [inline test/test_helper.ts:11]: Nit, how come we need this? Should be able to just call the function at the call site if it truly returns a promise.

**Review by jonathan-codaio — COMMENTED** [review]
Overall LGTM once heavily tested.

File: runtime/common/marshaling/index.ts [hunk 4]
```diff
@@ -173,12 +185,72 @@ export function marshalValue(val: any): MarshaledValue {
...
+  if (useUnsafeVersionCompatibilityHack) {
+    return maybeChangeWireVersionOnBase64EncodedV8SerializedData(serialized);
+  }
+
+  return serialized;
 }
 
 export function unmarshalValueFromString(marshaledValue: string): any {
-  return unmarshalValue(deserialize(marshaledValue));
+  if (marshaledValue?.startsWith('/')) {
```
Comment (huayang-codaio) [inline runtime/common/marshaling/index.ts:248]: if `marshaledValue` could be undefined, maybe update the parameter type as well.

File: runtime/common/marshaling/index.ts [hunk 5]
```diff
@@ -173,12 +185,72 @@ export function marshalValue(val: any): MarshaledValue {
...
+// This is a temporary hack for ease of node upgrades. It turns out node's documentation
+// is incorrect when it says v8.serialize is backwards compatible: moving from node 14/16
+// to node 18 changes the v8 serialize wire version from 13 to 15, which can't be read
+// by node 14 or 16. If we look at what actually changed in the wire format, it shouldn't
+// impact us as long as we don't try to let people serialize buffer views.
+// The changes were:
+// - https://chromium-review.googlesource.com/c/v8/v8/+/3417189 (support for shared strings)
+// - https://chromium-review.googlesource.com/c/v8/v8/+/3386802 (a change to buffer views)
+// The hacky temporary strategy here is to manually change the wire version back from 14 or 15
+// to 13 by decoding and re-encoding the first few bytes in the string.
```
Comment (huayang-codaio) [inline runtime/common/marshaling/index.ts:197]: amazing. thank you.

**Review by huayang-codaio — APPROVED** [review]
makes sense to me. 

do you plan to rebuild all (or selected) the pack bundles after releasing this change?

**Comment by huayang-codaio** [comment]
re testing: can use an adhoc.

File: test/marshaling_test.ts [hunk 6]
```diff
@@ -72,6 +83,52 @@ describe('Marshaling', () => {
     assert.deepEqual(transform(new SomeClass('hi')), {message: 'hi'});
   });
 
+  it('can write values readable on node 14 from node 18', async () => {
```
Comment (dweitzman-codaio) [inline test/marshaling_test.ts:86]: FYI, I've added a test case for the node18 backwards compatibility by hard-coding serialized output here. I manually ran just this one test case on node18 and it passed, showing that for at least these cases it's harmless to swap the wire version. I copied these test cases from above, but happy to add more if anything else seems worthwhile

File: test/test_helper.ts (hunk 3 above)
Comment (dweitzman-codaio) [reply test/test_helper.ts]: Oops, I should have used `assert.throws()`. Removed this

File: runtime/common/marshaling/index.ts (hunk 4 above)
Comment (dweitzman-codaio) [reply runtime/common/marshaling/index.ts]: I can get rid of the question mark here. The `undefined` I had encountered was actually a weird edge case where typed arrays would get dropped / turned into undefined but since those are officially not supported anymore I just dropped that test case

**Comment by dweitzman-codaio** [comment]
> makes sense to me.
> 
> do you plan to rebuild all (or selected) the pack bundles after releasing this change?

No. The wacky serialization backwards-compatibility hack is so we can avoid that:
- If old packs return old serialized data, that's fine: a server running a newer node version can still read old data (it's forwards-compatible)
- The server will write data to packs using the serialization hack in a 14-compatible format

The only problem would be if we have node14 lambdas still alive for a loooong time and we want to upgrade the server yet again and there's a v8.serialization change we can't hack our way around. AWS is deprecating lambdas with node 14 in November, so we'll probably get kicked off node14 by AWS and need to rebuild everything at node18 long before we run into serialization compatibility issues.

File: helpers/legacy_marshal/index.ts (hunk 2 above)
Comment (dweitzman-codaio) [reply helpers/legacy_marshal/index.ts:94]: Renamed to internalUnmarshalValueForAnyNodeVersion

**Review by chris-codaio — APPROVED** [review]
